docs/proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,106 @@
+ 
+### 2. Startegy for refactoring Kube-Controller-Manager 
+
+In order to create a 100% cloud independent controller manager, the idea is to split the controller-manager into two separate binaries. 
+
+1. Cloud dependent controller-manager binary
+2. Cloud independent controller-manager binary
+
+The cloud dependent binary will run those loops that rely on cloudprovider. The rest of the controllers will be run in the cloud independent controller manager. The decision to run entire controller loops, rather than only the very minute parts that rely on cloud provider makes implementation simple. If drilled down further into the controller loops, it would be significantly harder to disentangle, due to tight coupling of the cloud provider specific code with the rest of the data structures/functions within a controller.
+
+There are four controllers that rely on cloud provider specific code. These are node controller, service controller, route controller and attach detach controller. Copies of each of these controllers have been bundled them together into one binary. This binary registers itself as a controller, and runs the cloud specific controller loops with the user-agent named "external-controller-manager".
+
+RouteController and serviceController are entirely cloud specific. Therefore, it is really simple to run these two controller loops separately. 
+
+NodeController does a lot more than just talk to the cloud. Therefore, this can be split up into two different controllers. 
+
+Finally, The attachDetachController is tricky, and it is not simple to disentangle it from the controller-manager easily, therefore, this will be addressed with Flex Volumes (Discussed under a separate section below)
+
+### 3. Strategy for refactoring Kubelet 
+
+The majority of the calls by the kubelet to the cloud is done during the initialization of the Node Object. The other uses are for configuring Routes (in case of GCE), scrubbing DNS, and periodically polling for IP addresses. 
+
+All of the above steps, except the Node initialization step can be moved into a controller. Specifically, IP address polling, and configuration of Routes can be moved into the cloud dependent controller manager.
+
+Scrubbing DNS, after discussing with @thockin, was found to be redundant. So, it can be disregarded. It is being removed.
+
+Finally, Node initialization needs to be addressed. This is the trickiest part. Pods will be scheduled even on uninitialized nodes. This can lead to scheduling pods on incompatible zones, and other wierd errors. Therefore, an approach is needed where kubelet can create a Node, but mark it as "NotReady". Then, some asynchronous process can update it and mark it as ready. This is now possible because of the concept of ExternalAdmissionControllers introduced by @justinsb in these PRs - (https://github.com/kubernetes/kubernetes/pull/36210, https://github.com/kubernetes/kubernetes/pull/36209) 
```

Comment:
@anguslees I actually had this concern in the back of my head. I'm glad you brought about this issue, because we can now start discussion on how to approach this issue. 

As we start refactoring cloud providers, we could use a cloudprovider plugin (like cni) or start another service on each machine. I think one of these should solve your problem?